File: tmb-workshop-fhl/Day2/ex2v2.R
setwd("tmb-workshop-fhl/Day2")

require(TMB)

compile("ex2v2.cpp")
dyn.load(dynlib("ex2v2"))

################################################################################

nprey <- scan("EX2.dat",skip=1,n=1)
Ndata <- scan("EX2.dat",skip=3,n=1)
data <- read.table("EX2.dat",skip=4,header=TRUE,nrow=Ndata)
colnames(data) <- c("Predator", "Prey1", "Prey2",
                    "Prey3", "C1", "C2", "C3")

Model_type = 4

mod_dat <- list(Predator = data[,1],
                Prey = as.matrix(data[,2:4]),
                Consump = as.matrix(data[,5:7]))

# Add Model_type to our list of data
mod_dat$Model_type <- Model_type
mod_dat$Ndata <- Ndata

# Parameter list
parameters <- list(logalpha = rep(0, 3), logbeta = rep(0, 3), gamma = rep(2.1, 3),
                   loglambda = rep(0, 3))

# Create maps for different model types


if (Model_type==1) map <- list(logbeta = factor(rep(NA, 3)), gamma = factor(rep(NA, 3)), 
                               loglambda = factor(rep(NA, 3)))
if (Model_type==2) map <- list(gamma = factor(rep(NA, 3)), loglambda = factor(rep(NA, 3)))
if (Model_type==3) map <- list(loglambda = factor(rep(NA, 3)))
if (Model_type==4) map <- list(gamma = factor(rep(NA, 3)))
################################################################################

model <- MakeADFun(mod_dat, parameters, DLL="ex2v2",map=map)
fit <- nlminb(model$par, model$fn, model$gr)
best <- model$env$last.par.best
rep <- sdreport(model)
print(rep)

# Model 4 predictions
summary(rep)
Pred <- model$report()

#Predict consumption from 0 to 3.5
length(Pred$Pred)
plot(data$C1, data$Predator)
lines(data$C1, Pred$Pred)

plot(data$C2, data$Predator)
plot(data$C3, data$Predator)

File: tmb-workshop-fhl/Day2/ex2v2.cpp
#include <TMB.hpp>


template <class Type> Type square(Type x){return x*x;}


template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(Predator);
  DATA_MATRIX(Prey);
  DATA_MATRIX(Consump);
  DATA_INTEGER(Ndata);
  DATA_INTEGER(Model_type);

  PARAMETER_VECTOR(logalpha);
  PARAMETER_VECTOR(logbeta);
  PARAMETER_VECTOR(gamma);
  PARAMETER_VECTOR(loglambda);

  vector<Type> alpha = exp(logalpha);
  vector<Type> beta = exp(logbeta);
  vector<Type> lambda = exp(loglambda);

  matrix<Type> Pred(Ndata, 3);
  //vector<Type> PredY(3);
  
  //Type dummy = 2;

  Type neglogL = 0.0;

  if (Model_type == 1)
   {
    
    for(int obs = 0; obs < Ndata; obs++) 
    {
      for(int spp = 0; spp < 3; spp++)
      {
      Pred(obs, spp) = alpha(spp) * Predator(obs);
      }}
   }
	//for (II=0;II<=3;II++)
	// PredY(II-1) = Linf/(1+exp(-log(19)*(float(II)-a50)/Delta));
  // }
   
  if (Model_type == 2)
   {
    for(int obs = 0; obs < Ndata; obs++)
    {
      for(int spp = 0; spp < 3; spp++)
      {
    Pred(obs, spp) = (alpha(spp) * Predator(obs))/(1 + beta(spp) * Prey(obs, spp));
      }}
	//for (II=1;II<=20;II++)
	// PredY(II-1) = Linf*(1-exp(-Kappa*(float(II)-t0)));
   }
  
  if (Model_type == 3)
  {
    for(int obs = 0; obs < Ndata; obs++)
    {
      for(int spp = 0; spp < 3; spp++)
      {
    Pred(obs, spp) = (alpha(spp) * Predator(obs) * pow(Prey(obs, spp), gamma(spp) - 1)) / (1 + pow(beta(spp) * Prey(obs, spp), gamma(spp)));
      }}
  //for (II=1;II<=20;II++)
    // PredY(II-1) = Linf*(1-exp(-Kappa*(float(II)-t0)));
  }
  
  if (Model_type == 4)
  {
    for(int obs = 0; obs < Ndata; obs++)
    {
      for(int spp = 0; spp < 3; spp++)
      {
    Pred(obs, spp) = (alpha(spp) * Predator(obs)) / (1 + beta(spp) * Prey(obs, spp) + lambda(spp) * Predator(obs));
      }}
    //for (II=1;II<=20;II++)
    // PredY(II-1) = Linf*(1-exp(-Kappa*(float(II)-t0)));
  }
  
 for(int obs = 0; obs < Ndata; obs++) 
 {
  for(int spp = 0; spp < 3; spp++)
  {
    //neglogL += -sum(dnorm(log(Consump(obs,spp)), log(Pred(obs,spp)), 1, true));
    neglogL += pow(log(Consump(obs,spp)) - log(Pred(obs,spp)),2);
  //std::cout << alpha << " " << neglogL  << "\n";
  }}
 
  REPORT(Pred);
  ADREPORT(alpha);
  //REPORT(dummy);
  return neglogL;
}
